### PR TITLE
[Kernel] [V1] Fix performance regression for triton unified attention

### PR DESCRIPTION
**Pull request by tdoublep**
We have observed a pretty severe (~40%) performance regression for the `triton_unified_attention` kernel when moving from triton 3.2 to triton 3.3. 

After a lot of investigation, I was able to figure out that it came from [this](https://github.com/triton-lang/triton/pull/5512) commit to Triton. This change reworked the way that Triton determines what kernel arguments are constant. It seems that before this PR, Triton was (correctly) detecting that `stride_k_cache_3` and `stride_v_cache_3` are constant and leveraging this to obtain a faster kernel. For whatever reason, the new logic doesn't do this and we need to explicitly tell the compiler to interpret these strides as constant.

This minor change recovers the performance. 

There is one other small change: the TritonBackend no longer works on main because of [this](https://github.com/vllm-project/vllm/blob/main/vllm/v1/attention/backends/flash_attn.py#L287) assert. Since the TritonBackend re-uses the `FlashAttentionMetadata` and `FlashAttentionMetadataBuilder` directly (this is intentional to reduce duplicate code), this assert doesn't really make sense unless `TritonAttentionImpl` inherits from `FlashAttentionImpl`. Happy to consider other ways to solve that, but would like to avoid create entirely new `TritonAttentionMetadata` etc that are just duplicates. 

cc @SageMoore @bringlein 
  
Main branch:
```
++++++++++++++++++ Repetition 0 ++++++++++++++++++
============ Serving Benchmark Result ============
Successful requests:                     985       
Benchmark duration (s):                  26.18     
Total input tokens:                      209782    
Total generated tokens:                  125289    
Request throughput (req/s):              37.62     
Output token throughput (tok/s):         4785.69   
Total Token throughput (tok/s):          12798.78  
---------------Time to First Token----------------
Mean TTFT (ms):                          4889.40   
Median TTFT (ms):                        4844.86   
P99 TTFT (ms):                           8946.27   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          110.97    
Median TPOT (ms):                        53.41     
P99 TPOT (ms):                           390.86    
---------------Inter-token Latency----------------
Mean ITL (ms):                           46.36     
Median ITL (ms):                         28.78     
P99 ITL (ms):                            324.39    
==================================================

=++++++++++++++++++ Repetition 1 ++++++++++++++++++
============ Serving Benchmark Result ============
Successful requests:                     985       
Benchmark duration (s):                  24.18     
Total input tokens:                      209782    
Total generated tokens:                  125289    
Request throughput (req/s):              40.74     
Output token throughput (tok/s):         5181.56   
Total Token throughput (tok/s):          13857.49  
---------------Time to First Token----------------
Mean TTFT (ms):                          4455.20   
Median TTFT (ms):                        4315.81   
P99 TTFT (ms):                           8447.30   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          103.56    
Median TPOT (ms):                        49.74     
P99 TPOT (ms):                           312.85    
---------------Inter-token Latency----------------
Mean ITL (ms):                           43.30     
Median ITL (ms):                         26.91     
P99 ITL (ms):                            322.73    
==================================================

++++++++++++++++++ Repetition 2 ++++++++++++++++++
============ Serving Benchmark Result ============
Successful requests:                     985       
Benchmark duration (s):                  24.19     
Total input tokens:                      209782    
Total generated tokens:                  125305    
Request throughput (req/s):              40.73     
Output token throughput (tok/s):         5181.00   
Total Token throughput (tok/s):          13854.88  
---------------Time to First Token----------------
Mean TTFT (ms):                          4453.85   
Median TTFT (ms):                        4283.36   
P99 TTFT (ms):                           8430.35   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          103.24    
Median TPOT (ms):                        49.79     
P99 TPOT (ms):                           311.08    
---------------Inter-token Latency----------------
Mean ITL (ms):                           43.17     
Median ITL (ms):                         26.91     
P99 ITL (ms):                            325.18    
==================================================
```

With the changes from this PR:
```
++++++++++++++++++ Repetition 0 ++++++++++++++++++
============ Serving Benchmark Result ============
Successful requests:                     985       
Benchmark duration (s):                  18.36     
Total input tokens:                      209782    
Total generated tokens:                  125289    
Request throughput (req/s):              53.64     
Output token throughput (tok/s):         6823.19   
Total Token throughput (tok/s):          18247.83  
---------------Time to First Token----------------
Mean TTFT (ms):                          3535.61   
Median TTFT (ms):                        3447.89   
P99 TTFT (ms):                           6488.40   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          78.28     
Median TPOT (ms):                        38.20     
P99 TPOT (ms):                           231.30    
---------------Inter-token Latency----------------
Mean ITL (ms):                           32.68     
Median ITL (ms):                         20.25     
P99 ITL (ms):                            235.62    
==================================================

++++++++++++++++++ Repetition 1 ++++++++++++++++++
============ Serving Benchmark Result ============
Successful requests:                     985       
Benchmark duration (s):                  17.13     
Total input tokens:                      209782    
Total generated tokens:                  125289    
Request throughput (req/s):              57.50     
Output token throughput (tok/s):         7313.20   
Total Token throughput (tok/s):          19558.30  
---------------Time to First Token----------------
Mean TTFT (ms):                          3358.98   
Median TTFT (ms):                        3277.33   
P99 TTFT (ms):                           6301.26   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          76.65     
Median TPOT (ms):                        36.04     
P99 TPOT (ms):                           230.60    
---------------Inter-token Latency----------------
Mean ITL (ms):                           30.80     
Median ITL (ms):                         18.56     
P99 ITL (ms):                            232.08    
==================================================


++++++++++++++++++ Repetition 2 ++++++++++++++++++
============ Serving Benchmark Result ============
Successful requests:                     985       
Benchmark duration (s):                  17.37     
Total input tokens:                      209782    
Total generated tokens:                  125289    
Request throughput (req/s):              56.72     
Output token throughput (tok/s):         7214.31   
Total Token throughput (tok/s):          19293.85  
---------------Time to First Token----------------
Mean TTFT (ms):                          3571.61   
Median TTFT (ms):                        3417.45   
P99 TTFT (ms):                           6460.17   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          76.82     
Median TPOT (ms):                        36.22     
P99 TPOT (ms):                           231.26    
---------------Inter-token Latency----------------
Mean ITL (ms):                           30.74     
Median ITL (ms):                         18.33     
P99 ITL (ms):                            234.79    
==================================================
```

### CODE REVIEW

**Review by SageMoore — APPROVED** [review]

File: vllm/v1/attention/backends/triton_attn.py [hunk 1]
```diff
@@ -5,13 +5,13 @@
 import torch
 
 from vllm import _custom_ops as ops
-from vllm.attention.backends.abstract import (AttentionBackend, AttentionImpl,
+from vllm.attention.backends.abstract import (AttentionBackend,
```
Comment (robertgshaw2-redhat) [inline vllm/v1/attention/backends/triton_attn.py:8]: why is this change needed?

File: vllm/attention/ops/triton_unified_attention.py [hunk 2]
```diff
@@ -56,11 +56,11 @@ def kernel_unified_attention_2d(
     stride_k_cache_0: tl.int64,  # int
     stride_k_cache_1: tl.int64,  # int
     stride_k_cache_2: tl.int64,  # int
-    stride_k_cache_3: tl.int64,  # int
+    stride_k_cache_3: tl.constexpr,  # int
     stride_v_cache_0: tl.int64,  # int
     stride_v_cache_1: tl.int64,  # int
     stride_v_cache_2: tl.int64,  # int
-    stride_v_cache_3: tl.int64,  # int
+    stride_v_cache_3: tl.constexpr,  # int
```
Comment (mgoin) [inline vllm/attention/ops/triton_unified_attention.py:63]: Why not make all of the strides constexpr to be safe?

**Comment by mgoin** [comment]
Out of curiosity, are the benchmarks you reported CUDA or ROCm?

**Comment by LucasWilkinson** [comment]
> There is one other small change: the TritonBackend no longer works on main because of [this](https://github.com/vllm-project/vllm/blob/main/vllm/v1/attention/backends/flash_attn.py#L287) assert. Since the TritonBackend re-uses the FlashAttentionMetadata and FlashAttentionMetadataBuilder directly (this is intentional to reduce duplicate code), this assert doesn't really make sense unless TritonAttentionImpl inherits from FlashAttentionImpl. Happy to consider other ways to solve that, but would like to avoid create entirely new TritonAttentionMetadata etc that are just duplicates.

I discussed this with @bringlein we should avoid subclassing FlashAttention, the flash attention API is not as stable as people think and does change from time to time. We should be able update FlashAttention without needing to worry about a bunch of downstream attention backends (otherwise we've kinda defeated the point of having the attention backend abstraction). E.g. right now by using `FlashAttentionMetadataBuilder` directly when building the Triton metadata you are also calling the new FA ahead of time scheduler kernel only to ignore the results.

A temporary workaround could be to do:

```
class TritonAttentionMetadata(FlashAttentionMetadata):
    def __init__(self, runner: "GPUModelRunner", kv_cache_spec: AttentionSpec,
                 block_table: BlockTable):
        super().__init__(runner, kv_cache_spec, block_table)
        self.aot_schedule = False
```

this should fix the assert without having to subclass from `FlashAttentionImpl`


we should work on a new PR to have a operate Triton metadata and in the process move things like `make_local_attention_virtual_batches` into common area (like `utils.py` or `local_attn_common.py`), and maybe move `_get_sliding_window_configs` to a common area to and abstract so we can pass in the the impl class

File: vllm/attention/ops/triton_unified_attention.py (hunk 2 above)
Comment (LucasWilkinson) [reply vllm/attention/ops/triton_unified_attention.py:63]: I would think its best to not to avoid over recompilation, but for the last stride it makes sense since this is almost always 1 (and when it is 1 we want the compiler to optimize around this, i.e. use wider loads)

**Review by tlrmchlsmth — APPROVED** [review]
Nice find!

File: vllm/v1/attention/backends/triton_attn.py [hunk 3]
```diff
@@ -56,7 +56,7 @@ def get_builder_cls() -> type["FlashAttentionMetadataBuilder"]:
         return FlashAttentionMetadataBuilder
 
 
-class TritonAttentionImpl(AttentionImpl):
+class TritonAttentionImpl(FlashAttentionImpl):
```
Comment (LucasWilkinson) [inline vllm/v1/attention/backends/triton_attn.py:59]: I think we should do:
```
class TritonAttentionMetadata(FlashAttentionMetadata):
    def __init__(self, runner: "GPUModelRunner", kv_cache_spec: AttentionSpec,
                 block_table: BlockTable):
        super().__init__(runner, kv_cache_spec, block_table)
        self.aot_schedule = False
```
instead of this so we also avoid calls to the FA AOT scheduler unless you see an issue with this approach

**Review by LucasWilkinson — CHANGES_REQUESTED** [review]

File: vllm/attention/ops/triton_unified_attention.py (hunk 2 above)
Comment (tdoublep) [reply vllm/attention/ops/triton_unified_attention.py:63]: Avoiding recompilation is one reason, but I also think for really long sequences there is a risk that the strides can overflow unless they are explicitly marked as `tl.int64`. This can't happen for the `stride_k_cache_3` and `stride_v_cache_3` though, so I think we are safe to do this.

File: vllm/v1/attention/backends/triton_attn.py (hunk 3 above)
Comment (tdoublep) [reply vllm/v1/attention/backends/triton_attn.py:59]: Sure, trying this now. 

I will implement a clean version of `TritonAttentionMetadata` in a follow-on PR, I think they might be some benefits to the kernel if we do this.

Comment (tdoublep) [reply vllm/v1/attention/backends/triton_attn.py]: done

**Comment by tdoublep** [comment]
> Out of curiosity, are the benchmarks you reported CUDA or ROCm?

@mgoin Benchmarks above are on H100, I believe @bringlein has also confirmed this behaviour on MI300x too.

File: vllm/v1/attention/backends/triton_attn.py (hunk 3 above)
Comment (tlrmchlsmth) [reply vllm/v1/attention/backends/triton_attn.py]: @LucasWilkinson nice catch. Is it good to go now?

Comment (bringlein) [reply vllm/v1/attention/backends/triton_attn.py]: I similar workaround is done in the PR https://github.com/vllm-project/vllm/pull/16606

**Review by LucasWilkinson — APPROVED** [review]
LGTM, thanks for updating!

**Comment by tdoublep** [comment]
V1 test failures look unrelated (something to do with Eagle)

**Comment by ekagra-ranjan** [comment]
@tdoublep - Nice job!
which model was this benchmarked on in the PR description? Does it mean all model which use Flash attention will see 40% due to this PR?
